README.md
# challenge-data-processing

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar challenge-data-processing-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-03-13
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-03-13
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/challenge-data-processing/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/challenge-data-processing/compare/0.1.0...0.1.1


test/challenge_data_processing/core_test.clj
(ns challenge-data-processing.core-test
  (:require
    [clojure.core.match :refer [match]]
    [clojure.java.io :as io]
    [clojure.set :as set]
    [clojure.string :as string]
    [clojure.test :refer [deftest testing is]]
    [clojure.walk :as walk]
    [jsonista.core :as json]))


(defonce lipstick (-> "lipstick.json" io/resource slurp json/read-value))


(defn csv
  [csv-head csv-body]
  (->> (cons csv-head csv-body)
       (map #(string/join "," %))
       (string/join (System/lineSeparator))))


(defn sort-by-price-asc
  [list-items]
  (into (sorted-set-by (fn [{price-01 "price"} {price-02 "price"}]
                         (<= (Double/parseDouble price-01) (Double/parseDouble price-02))))
        list-items))


(defn purchase-min-price
  [{:keys [balance list-items purchased-items]}]
  (let [[{:strs [brandName price] :as item} & other-items] (sort-by-price-asc list-items)]
    (if (and price (< 0 balance))
      {:balance (- balance (Double/parseDouble price))
       :list-items (remove #(-> % (get "brandName") (= brandName)) other-items)
       :purchased-items (conj purchased-items item)}
      {:balance -1
       :list-items []
       :purchased-items purchased-items})))


(defn purchase-item-possible?
  [{:keys [balance]}]
  (<= 0 balance))


(defn purchased-items-maximum-possible
  [balance list-items]
  (->> {:balance balance :list-items list-items}
       (iterate purchase-min-price)
       (take-while purchase-item-possible?)
       (last)
       (:purchased-items)))


(defn purchased-items-csv-body
  [list-items]
  (for [{:strs [itemId brandName price]} list-items]
    [(str "'" itemId "'")
     (str "'" brandName "'")
     (str "\"" price "\"")]))


(deftest question-05-test
  (let [list-items [{"price" "10" "brandName" "a"}
                    {"price" "5" "brandName" "b"}
                    {"price" "15" "brandName" "b"}
                    {"price" "5" "brandName" "c"}]]
    (is (= [{:balance 5
             :list-items list-items}
            {:balance 0.0
             :list-items [{"price" "5" "brandName" "b"}
                          {"price" "10" "brandName" "a"}
                          {"price" "15" "brandName" "b"}]
             :purchased-items [{"price" "5" "brandName" "c"}]}]
           (->> {:balance 5 :list-items list-items}
                (iterate purchase-min-price)
                (take-while purchase-item-possible?))))
    (is (= [{:balance 10
             :list-items list-items}
            {:balance 5.0
             :list-items [{"price" "5" "brandName" "b"}
                          {"price" "10" "brandName" "a"}
                          {"price" "15" "brandName" "b"}]
             :purchased-items [{"price" "5" "brandName" "c"}]}
            {:balance 0.0
             :list-items [{"price" "10" "brandName" "a"}]
             :purchased-items [{"price" "5" "brandName" "b"} {"price" "5" "brandName" "c"}]}]
           (->> {:balance 10 :list-items list-items}
                (iterate purchase-min-price)
                (take-while purchase-item-possible?))))
    (is (= [{:balance 50
             :list-items list-items}
            {:balance 45.0
             :list-items [{"price" "5" "brandName" "b"}
                          {"price" "10" "brandName" "a"}
                          {"price" "15" "brandName" "b"}]
             :purchased-items [{"price" "5" "brandName" "c"}]}
            {:balance 40.0
             :list-items [{"price" "10" "brandName" "a"}]
             :purchased-items [{"price" "5" "brandName" "b"} {"price" "5" "brandName" "c"}]}
            {:balance 30.0
             :list-items []
             :purchased-items [{"price" "10" "brandName" "a"} {"price" "5" "brandName" "b"} {"price" "5" "brandName" "c"}]}]
           (->> {:balance 50 :list-items list-items}
                (iterate purchase-min-price)
                (take-while purchase-item-possible?))))))


(deftest question-05-02-test
  (let [list-items (-> lipstick
                       (get-in ["mods" "listItems"]))
        result (->> (purchased-items-maximum-possible 250 list-items)
                    purchased-items-csv-body)]
    (is (= [["'224295407'" "'OEM'" "\"39.00\""]
            ["'341864493'" "'Novo'" "\"39.00\""]
            ["'331292065'" "'IMAGES'" "\"29.00\""]
            ["'350478554'" "'ZIRANMI'" "\"29.00\""]
            ["'346252619'" "'PEINIFEN'" "\"25.00\""]
            ["'278125439'" "'sivanna'" "\"24.00\""]
            ["'523686357'" "'Kiss Beauty'" "\"20.00\""]
            ["'316046847'" "'MENOW'" "\"19.00\""]
            ["'543952820'" "'Tanako'" "\"15.00\""]
            ["'502436001'" "'No Brand'" "\"7.00\""]]
           result))))


(comment
  (def lipstick (-> "lipstick.json" io/resource slurp json/read-value))
  (keys lipstick)
  ; question 01
  (->> (get-in lipstick ["mods" "listItems"])
       count-skus-csv-body
       (csv ["productUrl" "price" "originalPrice" "numberOfSKUs"])
       (spit "target/first.csv"))
  ; question 02
  (->> (get-in lipstick ["mods" "listItems"])
       filtered-price
       avg-price
       (spit "target/second.csv"))
  ; question 03
  (->> (get-in lipstick ["mods" "listItems"])
       count-products-by-brand-name
       json/write-value-as-string
       (spit "target/third.json"))
  ; question 04
  (->> lipstick
       image-vals
       file-names
       (string/join (System/lineSeparator))
       (spit "target/fourth.csv"))
  ; question 05
  (->> (get-in lipstick ["mods" "listItems"])
       (purchased-items-maximum-possible 250)
       purchased-items-csv-body
       (csv ["itemId" "brandName" "price"])
       (spit "target/fifth.csv"))
  (kaocha.repl/run-all))


(kaocha.repl/run-all)


test/challenge_data_processing/question_01_test.clj
(ns challenge-data-processing.question-01-test
  (:require
    [challenge-data-processing.csv :as csv]
    [challenge-data-processing.question-01 :as question-01]
    [clojure.java.io :as io]
    [clojure.test :refer [deftest testing is]]
    [jsonista.core :as json]))


(def lipstick (-> "lipstick.json" io/resource slurp json/read-value))


(deftest count-skus-csv-body-test
  (testing "Given a list of items,"
    (let [list-items (-> lipstick
                         (get-in ["mods" "listItems"]))]
      (testing "it should have the same count as original list"
        (is (= 100 (count (question-01/count-skus-csv-body list-items)))))

      (testing "it should count the number of skus"
        (is (= ["'//www.lazada.co.th/products/qianxiu-q127-the-new-moisturizing-and-waterproof-moisturizer-is-not-easy-to-wear-cokkicosmetic-i224295407-s946252295.html?search=1'"
                "\"39.00\""
                "'109.00'"
                6]
               (-> (question-01/count-skus-csv-body list-items) (first)))))

      (testing "it can be formatted as csv"
        (is (= "productUrl,price,originalPrice,numberOfSKUs\n'//www.lazada.co.th/products/qianxiu-q127-the-new-moisturizing-and-waterproof-moisturizer-is-not-easy-to-wear-cokkicosmetic-i224295407-s946252295.html?search=1',\"39.00\",'109.00',6"
               (->> (question-01/count-skus-csv-body list-items)
                    (take 1)
                    (csv/csv ["productUrl" "price" "originalPrice" "numberOfSKUs"]))))))))


test/challenge_data_processing/question_02_test.clj
(ns challenge-data-processing.question-02-test
  (:require
    [challenge-data-processing.question-02 :as question-02]
    [clojure.java.io :as io]
    [clojure.test :refer [deftest testing is]]
    [jsonista.core :as json]))


(def lipstick (-> "lipstick.json" io/resource slurp json/read-value))


(deftest question-02-test
  (testing "Given a list of items,"
    (let [list-items (-> lipstick
                         (get-in ["mods" "listItems"]))]
      (testing "filtered-price should return prices of items which has OEM brandName and number of skus more than 2"
        (is (= [39.0 59.0 59.0 59.0 49.0]
               (question-02/filtered-price list-items))))

      (testing "avg-price should compute the average"
        (is (= 53.0
               (-> list-items
                   question-02/filtered-price
                   question-02/avg-price)))))))


src/challenge_data_processing/core.clj
(ns challenge-data-processing.core
  (:gen-class)
  (:require
    [challenge-data-processing.question-01 :as question-01]
    [challenge-data-processing.question-02 :as question-02]
    [challenge-data-processing.question-03 :as question-03]
    [challenge-data-processing.question-04 :as question-04]
    [challenge-data-processing.question-05 :as question-05]
    [jsonista.core :as json]))


(defn -main
  [& _]
  (try
    (let [input (-> "lipstick.json"
                    slurp
                    json/read-value)]
      (println "Write answer to first.csv ...")
      (spit "first.csv" (question-01/answer input))
      (println "Write answer to second.csv ...")
      (spit "second.csv" (question-02/answer input))
      (println "Write answer to third.json ...")
      (spit "third.json" (question-03/answer input))
      (println "Write answer to fourth.csv ...")
      (spit "fourth.csv" (question-04/answer input))
      (println "Write answer to fifth.csv ...")
      (spit "fifth.csv" (question-05/answer input))
      (println "Done."))
    (catch Throwable ex
      (println ex))))


src/challenge_data_processing/question_04.clj
(ns challenge-data-processing.question-04
  (:require
    [clojure.core.match :refer [match]]
    [clojure.set :as set]
    [clojure.string :as string]
    [clojure.walk :as walk]))


(defn reduce-image-vals
  [seq-node]
  (->> seq-node
       (filter map?)
       (apply merge)
       keys
       (filter map?)
       (keep :image-vals)
       flatten
       (apply set/union)))


(defn merge-image-vals
  [map-node]
  (->> map-node
       keys
       (filter map?)
       (apply merge-with set/union)))


(defn find-image-vals
  [node]
  (let [type-node (when (map-entry? node)
                    :map-entry)]
    (match [type-node node]
      [:map-entry ["image" (image-val :guard string?)]] [{:image-vals #{image-val}} nil]
      [:map-entry [_ ([& seq-node] :seq)]] [{:image-vals (reduce-image-vals seq-node)} nil]
      [:map-entry [_ (map-node :guard map?)]] [(merge-image-vals map-node) nil]
      [:map-entry _] nil
      :else node)))


(defn image-vals
  [lipstick]
  (->> lipstick
       (walk/postwalk find-image-vals)
       keys
       (keep :image-vals)
       (apply set/union)))


(defn file-names
  [image-vals]
  (->> image-vals
       (map #(string/split % #"/"))
       (map last)
       (into #{})))


(defn answer
  [input]
  (->> input
       image-vals
       file-names
       (string/join (System/lineSeparator))))


src/challenge_data_processing/question_05.clj
(ns challenge-data-processing.question-05
  (:require
    [challenge-data-processing.csv :as csv]))


(defn sort-by-price-asc
  [list-items]
  (into (sorted-set-by (fn [{price-01 "price"} {price-02 "price"}]
                         (<= (Double/parseDouble price-01) (Double/parseDouble price-02))))
        list-items))


(defn purchase-min-price
  [{:keys [balance list-items purchased-items]}]
  (let [[{:strs [brandName price] :as item} & other-items] (sort-by-price-asc list-items)]
    (if (and price (< 0 balance))
      {:balance (- balance (Double/parseDouble price))
       :list-items (remove #(-> % (get "brandName") (= brandName)) other-items)
       :purchased-items (conj purchased-items item)}
      {:balance -1
       :list-items []
       :purchased-items purchased-items})))


(defn purchase-item-possible?
  [{:keys [balance]}]
  (<= 0 balance))


(defn purchased-items-maximum-possible
  [balance list-items]
  (->> {:balance balance :list-items list-items}
       (iterate purchase-min-price)
       (take-while purchase-item-possible?)
       (last)
       (:purchased-items)))


(defn purchased-items-csv-body
  [list-items]
  (for [{:strs [itemId brandName price]} list-items]
    [(str "'" itemId "'")
     (str "'" brandName "'")
     (str "\"" price "\"")]))


(defn answer
  [input]
  (->> (get-in input ["mods" "listItems"])
       (purchased-items-maximum-possible 250)
       purchased-items-csv-body
       (csv/csv ["itemId" "brandName" "price"])))
